DateAdd and DateSubtract expression

qnkhuat opened the pull request:
Implement `:date-add` and `:date-subtract` methods as spec'ed in this [project doc](https://www.notion.so/metabase/Unlock-advanced-date-time-manipulation-038f5cf80f364d7cb93341784169743c)

date-add
<img width="419" alt="Screen Shot 2022-09-20 at 23 16 58" src="https://user-images.githubusercontent.com/25661381/191311631-a662cbd2-a0af-455f-9aad-baa21012ef18.png">


date-subtract
<img width="418" alt="Screen Shot 2022-09-20 at 23 17 13" src="https://user-images.githubusercontent.com/25661381/191311655-ba68ec55-a1ea-479a-9c34-5d2b25d9e53d.png">

jeff303 committed:
Implement advanced date/time/zone manipulation, part 1

WORK IN PROGRESS

Incorporate new functions into MBQL and add tests:
 - get-year
 - get-quarter
 - get-month
 - get-day
 - get-day-of-week
 - get-hour
 - get-minute
 - get-second

jeff303 committed:
Fix BigQuery implementations to call extract

Mark as not supported in legacy driver

jeff303 committed:
Merge branch 'master' into advanced-datetime-handling-1-extraction-functions

jeff303 committed:
Add date extraction fns for Postgres

jeff303 committed:
Disable in MongoDB (for now at least)

Disable in BigQuery (legacy driver)

Add implementations for presto-jdbc

jeff303 committed:
Merge branch 'master' into advanced-datetime-handling-1-extraction-functions

qnkhuat committed:
merge with master

qnkhuat committed:
Misc cleanup from Jeff's PR

qnkhuat committed:
Update Jeff's implementation of bigquery-cloud-sqk

qnkhuat committed:
Reorganized tests

qnkhuat committed:
Mongo

qnkhuat committed:
Oracle

qnkhuat committed:
Sqlserver

qnkhuat committed:
Sqlite

qnkhuat committed:
Add casting supports for presto

qnkhuat committed:
Remove Jeff's implementation of presto-jdbc because  its parent is
sql-jdbc

qnkhuat committed:
Update presto-jdbc tests to use the same catalog for all datasets

qnkhuat committed:
remove unused bindings, requires

qnkhuat committed:
we should be able to use date extract functions in summarization

qnkhuat committed:
fix linting

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat committed:
test smarter, not longer

qnkhuat committed:
first stab

qnkhuat committed:
making vertica works

qnkhuat committed:
should fix tests for presto and sparksql

qnkhuat committed:
Debugging bigquery

qnkhuat committed:
bigquery fix and implement for mongo + tests

qnkhuat committed:
date arithmetic only supports for mongo 4>

qnkhuat committed:
fix formatting for presto tests

qnkhuat committed:
more tests for nested usecase and fix for bigquery

qnkhuat committed:
remove duplicated test

ranquild committed:
Add date extraction functions to the expression editor (#25382)

qnkhuat committed:
make sure the semantic type of aggregated columns are integer

qnkhuat committed:
requires mbql.s

qnkhuat committed:
Merge branch 'ngoc-advanced-datetime-extraction' into ngoc-add-substract-time

qnkhuat committed:
update schema

qnkhuat committed:
implement date-add and date-subtract for Fe + normalize the unit in BE

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-functions) :mongo)
```
Use `(testing ...)` to name the test cases.

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-functions) :mongo)
+      (doseq [[col-type field-id] [[:datetime (mt/id :times :dt)] [:text-as-datetime (mt/id :times :as_dt)]]]
+        (doseq [op [:date-add :date-subtract]]
+          (doseq [unit [:year :quarter :month :day :hour :minute :second]]
+            (doseq [[expected query]
```
Collapse these into a single `doseq` and save some horizontal space.

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+                 amount
+                 (- amount))
+        result (cond
+                 ;; unlike everybody else, vertica decided when adding 1 year it will do that by adding 365 days,
+                 ;; or 90 days if you want to add a quarter.
+                 ;; That means if you do `select date '2004-02-02' + interval '2 year';` in vertica, it will returns
+                 ;; `2006-02-01` because 2004 is a leap year. All others DBs will return `2006-02-02` instead.
```
Don't we want to eliminate this difference? (I could argue pro and contra, so I'm just wondering.)

qnkhuat replied:
https://github.com/metabase/metabase/pull/25414/commits/9d32c7b56af1f55b292678d582d0e3183f7078f4

I make the behavior of vertica consistent with other dbs in this commit. WDYT?

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+        (doseq [op [:date-add :date-subtract]]
+          (doseq [unit [:year :quarter :month :day :hour :minute :second]]
+            (doseq [[expected query]
+                    [[[[(date-math op #t "2004-03-19 09:19:09" 2 unit col-type)] [(date-math op #t "2008-06-20 10:20:10" 2 unit col-type)]
+                       [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)] [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]]
```
I wonder if it was easier to follow if the field values were read from the database instead of being hardwired here.

metamben commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+                 (is (= (set expected) (set (test-date-extract query))))))))))
+
+     (when-not (> (mongo-major-version (mt/db)) 4)
+       (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Date arithmetic not supported in versions before 5"
+                             (mt/compile (mt/mbql-query times {:expressions {"expr" [:date-add [:field (mt/id :times :dt) nil] 2 :year]}
+                                                               :fields      [[:expression "expr"]]}))))
+       (is (thrown-with-msg? clojure.lang.ExceptionInfo #"Date arithmetic not supported in versions before 5"
+                             (mt/compile (mt/mbql-query times {:expressions {"expr" [:date-subtract [:field (mt/id :times :dt) nil] 2 :year]}
+                                                               :fields      [[:expression "expr"]]}))))))))
```
You could `doseq` these two as well.

metamben reviewed (commented):
Looks reasonable, but I can't really judge the completeness. I hope someone else will have a look at it too.

qnkhuat commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -594,34 +594,49 @@
+  unit     ArithmeticDateTimeUnit)
+
+(defclause ^{:requires-features #{:date-functions}} date-subtract
+  datetime StringExpressionArg,
```
ah right, should be ExpressionArg to match other extract functions.

While we're on this, do you think we need a whole expression for datetime?

ranquild reviewed (approved):
FE 👍

dpsutton commented on modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:
```diff
@@ -174,8 +182,10 @@
+  [{base-type :base_type, effective-type :effective_type, database-type :database_type,
+    coercion-strategy :coercion_strategy}]
   (or (database-type->temporal-type database-type)
+      (coercion-strategy->temporat-type coercion-strategy)
```
I'm surprised this is necessary. When setting a corecion_strategy the effective_type should be set to the resulting type. (eg, base type is string, coercion is Coercion/ISO8601->DateTime, and effective is :type/DateTime or equivalent). The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

qnkhuat replied:
> The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

You're right, and the bug is the mapping between our type system and the actual type we cast is mismatched for bigquery.

Specifically when you do a `:Coercion/ISO8601->DateTime` we do 2 things:
- we set the effective type of this column to `type/DateTime`
- we cast this column to `timestamp` ([ref](https://github.com/metabase/metabase/blob/a99ec1dfbb5bccaed4c827054df738789a0a0ea1/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L728))

The mismatch here is we interpret `type/Datetime` as `datetime` type, but the actually db-type is `timestamp` (because we cast to timestamp).

This is a problem now because bigquery distinguishes between `timestamp` and `datetime` when you do datetime arithmetic.
See [this](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L721) comment for details. 

--- 

I think the correct fix here is to make sure the effective type for `:Coercion/ISO8601->DateTime` is `type/DateTimeWithTZ`, this way we will interpret it correctly as `timestamp`.

But the [code](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/shared/src/metabase/types.cljc#L355) for this is not dispatched per driver, And considering how a big `radius` this change affect our codebase, I think it's better to solve it within bigquery like I did here.

WDYT?

qnkhuat replied:
actually, we could just make `:Coercion/ISO8601->DateTime` cast to `datetime` for bigquery.
Any reason this is a no-go?

qnkhuat replied:
I went ahead with making `:Coercion/ISO8601->DateTime` to cast to `datetime` in this commit https://github.com/metabase/metabase/pull/25414/commits/786970c24f11c2ea8fe5277e42b9fa80b23a798e

dpsutton commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -132,3 +142,140 @@
+    (t/format fmt result)))
+
+(defn- mongo-major-version [db]
+  (-> (get-in db [:details :version])
```
I think this should be made resilient to the version string not being present. The `str/split` and `parse-long` will both error out if that version string is missing. If we do want to error we should have a better error message than those will provide.

camsaul replied:
I don't really love the pattern of checking the MongoDB version here. If supporting this stuff is dependent on the Mongo version, we should be using `driver/database-supports?` to check support for the feature rather than `driver/supports?`. That being said we don't have test tooling around that yet -- `mt/normal-drivers-with-feature` checks `driver/supports?` rather than `driver/database-supports?`. So I won't put my foot down on this.

That being said if you want to do this the right way you can always do something like

```clj
(mt/test-drivers (mt/normal-drivers-with-feature :date-arithmetics)
  (when (driver/database-supports? driver/*driver* (mt/db) :date-arithmetics)
    ...))
```

Better yet, you can turn that into a test util macro so we can start using that everywhere.

qnkhuat committed:
nest doseqs

qnkhuat committed:
StringExpressionArg -> ExpressiongArg for date-add and date-subtract
schema

qnkhuat committed:
Bigquery: cast Coercion/ISO8601->DateTime instead of infer coercion type

qnkhuat committed:
Vertica Use timestampadd for date arithmetics

qnkhuat commented:
Hi @nllho. could you look into the docs for these 2 news `date-add` and `date-subtract` expressions? I have screenshots of it in the PR's description.

qnkhuat committed:
merge with upstream

nllho commented:
@qnkhuat Here are my suggestions :)

I changed "datetime" to "date or timestamp value" because `datetime` is a specific data type in some databases.

But if the function _only_ works with specific data types (`datetime`, `timestamp`, `date`, etc.), then feel free to replace "date or timestamp value" with whatever is correct!

```
dateAdd(column, amount, unit)

Adds some amount of time to a date or timestamp value.

column

The name of the column with your date or timestamp values.

amount

The number of time units to be added.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".

```

```
dateSubtract(column, amount, unit)

Subtracts some amount of time from a date or timestamp value.

column

The name of the column with your date or timestamp values.

amount

The number of time units to be subtracted.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".
```

qnkhuat commented:
@nllho Thanks for look into this. I have some feedback.
> Adds some amount of time to a date or timestamp value.

I'm concerned about the `amount of time` part, this _might_ suggest that this only work with time(hour, minute, second) but it actually work with date  unit too (day, month ,quarter, year)

> The name of the column with your date or timestamp values.

I'd avoid using `the name of the column`  because it's technically incomplete. The `column` could be the result of an expression that returns a datetime column. Think case where we do `dateAdd(convertTimezone([created at], "Asia/Tokyo"), -1, "month")`

FWIW my version took some inspiration from this doc : https://www.techonthenet.com/excel/formulas/dateadd.php

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

nllho commented:
@qnkhuat Good points! I think we just want to make it clear that "units" refers to the "unit" parameter, which tells people the size of the time units that they can choose from.

Here's an updated suggestion :)

```
dateAdd(column, amount, unit)

Adds some units of time to a date or timestamp value.

column

The column with your date or timestamp values.

amount

The number of units to be added.

unit

"year", "month", "quarter", "day", "hour", "minute", or "second".
```

metamben reviewed (approved):
LGTM

qnkhuat committed:
update UI docs

qnkhuat commented:
Updated! Thank you @nllho

qnkhuat committed:
reverse Jeff's change to test/data/interface

qnkhuat committed:
nested expression and filter using date extraction fns

qnkhuat committed:
rename

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat committed:
more tests and rename the schema

qnkhuat committed:
add docstring

qnkhuat committed:
no recursive call in annotate for date-extract func

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -638,8 +646,23 @@
 
+(def ^:private ArithmeticDateTimeUnit
+  (s/named
+   (apply s/enum #{:default :second :minute :hour :day :week :month :quarter :year})
```
1. Should `:millisecond` be in here as well? Since some of the code above is written to support it.
2. I think we should disallow `:default` here. How does one add a `1 :default`? It doesn't make sense. We should require an explicit unit. When we use `:default` in other places it basically means "use this column at it's current resolution, don't apply truncation/bucketing"... doesn't really apply here.

```suggestion
   (apply s/enum #{:millisecond :second :minute :hour :day :week :month :quarter :year})
```

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -638,8 +646,23 @@
+   "arithmetic-datetime-unit"))
+
+(defclause ^{:requires-features #{:date-arithmetics}} date-add
+  datetime DateTimeExpressionArg,
```
```suggestion
  datetime DateTimeExpressionArg
```

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -638,8 +646,23 @@
+  unit     ArithmeticDateTimeUnit)
+
+(defclause ^{:requires-features #{:date-arithmetics}} date-subtract
+  datetime DateTimeExpressionArg,
```
```suggestion
  datetime DateTimeExpressionArg
```

camsaul commented on shared/src/metabase/mbql/schema.cljc:
```diff
@@ -833,6 +856,14 @@
    (partial is-clause? :case)                        case
    :else                                             Field))
 
+#_((s/validator DatetimeExpression)
+   [:date-add [:date-add [:field 3 nil] 2 :year] 2 :month])
+
+
+#_((s/validator ArithmeticExpression)
+   [:+ [:+ [:field 3 nil] 2] 2])
+
```
I'd personally rather not leave a bunch of comments for testing stuff all over the codebase. Take this and write an actual test

```suggestion
```

qnkhuat replied:
Oh I forgot to delete these from debugging. Will move.

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -1,9 +1,18 @@
             [metabase.test :as mt]
             [metabase.util.date-2 :as u.date]))
 
+(defn- formatting [x]
+  (if (number? x)
+    (int x)
+    (-> x
+        (str/replace  #"T" " ")
+        (str/replace  #"Z" ""))))
```
Can you give this a better name? `(formatting x)` is meaningless. `(format-timestamp x)` is much better, but you can probably do something better than that. Maybe `normalize-timestamp-str` or something.

I would also not use this for anything but timestamp strings. You should know which columns are which in the tests... use `int` for the ID column and `normalize-timestamp-str` for the timestamp columns.

```suggestion
(defn- normalize-timestamp-str [x]
  (-> x
      (str/replace  #"T" " ")
      (str/replace  #"Z" "")))
```

camsaul replied:
You probably don't even want to use this in `test-date-extract` anyway, right? I don't understand why you changed **that** test

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+          (testing (format "%s %s function works as expected on %s column for driver %s" op unit col-type driver/*driver*)
+            (is (= (set expected) (set (test-date-extract query))))))))
+
+   (mt/test-driver :mongo
+     ;; date arithmetic doesn't supports until mongo 5+
+     (when (> (mongo-major-version (mt/db)) 4)
+       (testing "date arithmetic with datetime columns"
+         (let [[col-type field-id] [:datetime (mt/id :times :dt)]]
+           (doseq [op               [:date-add :date-subtract]
+                   unit             [:year :quarter :month :day :hour :minute :second]
+                   [expected query]
+                   [[[[(date-math op #t "2004-03-19 09:19:09" 2 unit col-type)] [(date-math op #t "2008-06-20 10:20:10" 2 unit col-type)]
+                      [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)] [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]]
+                     {:expressions {"expr" [op [:field field-id nil] 2 unit]}
+                      :fields      [[:expression "expr"]]}]
+
+                    [(into [] (frequencies
+                                [(date-math op #t "2004-03-19 09:19:09" 2 unit col-type) (date-math op #t "2008-06-20 10:20:10" 2 unit col-type)
+                                 (date-math op #t "2012-11-21 11:21:11" 2 unit col-type) (date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]))
+                     {:expressions {"expr" [op [:field field-id nil] 2 unit]}
+                      :aggregation [[:count]]
+                      :breakout    [[:expression "expr"]]}]]]
+             (testing (format "%s %s function works as expected on %s column for driver %s" op unit col-type driver/*driver*)
+               (is (= (set expected) (set (test-date-extract query))))))))
```
Mongo-specific tests belong somewhere `modules/drivers/mongo/test`, not in the general tests. Make some of the stuff here public so you can reuse it there if you need to.

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-arithmetics) :mongo)
+      (testing "date arithmetic with datetime columns"
+        (doseq [[col-type field-id] [[:datetime (mt/id :times :dt)] [:text-as-datetime (mt/id :times :as_dt)]]
```
Suggestions:

Could you make column type part of the `doseq` as well? That way you have one mega `doseq` instead of three

These sorts of big `doseq`-based tests get sort of hard to read I think. Since a lot of the stuff below is declarative I'd consider moving what you can out into a top-level `def` and call it `date-math-test-cases` or something. Then the actual business logic can go in the test itself. It makes things a little easier on the eyes I think.

Just remember that you shouldn't be loading the dataset or calling `mt/id` in a top-level form... so your `def` should have column name (`:dt`) and the test itself can use that to get the actual Field

camsaul replied:
For the query you could change it to be a function maybe e.g. `:query-fn (fn [field-id] ...)`. IDK. Or just have it look at a dynamic var. Doesn't really matter. I just want this test to be easier to read

qnkhuat replied:
If we want to move this to a top-level def, then the expected need to be an `fn` too, because the `expected` are being generated dynamically.

So it'll probably look something like

```clojure
{:expected-fn (fn [col-type op unit col-type]... )
 :query-fn    (fn [field-id] ...))
```
I'm not sure this makes it easier to reason about the tests tho.

But I'm all ears on how to improve these tests, I've been going back and forth with it several times and still not fully satisfied with it.

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+                                                                  :fields      [[:expression "expr"]]}))))))))))
+
+(deftest date-math-with-extract-test
+  (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-arithmetics) :mongo)
```
If you're excluding a driver you should leave a comment with a note as to why you are doing so. If it's because there is a different version of the test for mongo then include a `[[link]]` to it in your comment

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-arithmetics) :mongo)
```
you gotta write a note about why we're skipping Mongo here.

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+                op                  [:date-add :date-subtract]
+                unit                [:year :quarter :month :day :hour :minute :second]
+                [expected query]
+                [[[[(date-math op #t "2004-03-19 09:19:09" 2 unit col-type)] [(date-math op #t "2008-06-20 10:20:10" 2 unit col-type)]
+                   [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)] [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]]
+                  {:expressions {"expr" [op [:field field-id nil] 2 unit]}
+                   :fields      [[:expression "expr"]]}]
```
I know tuples like `[expected query]` are more convenient than maps but seeing `[[[[` makes my eyes bleed... I think 

```clj
{:expected ...
 :query ...}
```

would be a bit more readable. Maybe if you move this stuff into a top-level `def` you can do that instead.

camsaul commented on test/metabase/query_processor_test/date_time_zone_functions_test.clj:
```diff
@@ -169,3 +177,141 @@
+                 "yyyy-MM-dd HH:mm:ss")]
+    (t/format fmt (u.date/add x unit amount))))
+
+(defn- mongo-major-version [db]
```
```suggestion
(defn- db-major-version [db]
```

Is this Mongo DB specific at all?

camsaul reviewed (approved):
Overall it looks good to me but I left a handful of notes about things. To summarize my suggestions:

- [x] Should we explicitly support `:millisecond`? I would vote yes just for completeness.
- [x] I think adding or subtracting a `:default` makes no sense. You should probably take that out
- [x] If you skip tests for a specific driver you have to put a comment in there explaining why. If it's because there's a different version of the test somewhere else then include a `[[link]]` to it in the comments. Ideally, you wouldn't skip tests for Mongo at all... maybe if you do the `driver/database-supports?` thing I suggested you wouldn't need to skip it for Mongo
- [x] Driver-specific tests should go in driver test namespaces. Ideally there should be zero occurrences of the `:mongo` keyword in `test/` at all. I know that's not the reality but it's something we should stive for
- [x] **Tests** should not be checking database version... that's a clear sign we need to implement `driver/database-supports?` rather than `driver/supports?`. Please implement it for MongoDB and have that method check the major version, and then update your test to call that as I suggested
- [x] The main test is very big and a little hard to read. See if there's a way to simplify it somehow, maybe by taking stuff like `:expected`/`:query` out and putting it in a top-level `def`.

Since all of those things are relatively minor I'm going to conditionally approve this PR so we you don't need to wait for another review before getting this in. So once you address my feedback feel free to merge.

qnkhuat committed:
- add supports for millisecond
- move mongo tests to mongo ns
- make driver/supports :date-arithmetics to check for mongo version

qnkhuat committed:
Tetss date-math for mongo

qnkhuat committed:
get-unit -> datetime-extract(column, unit)

qnkhuat committed:
desguar nested datetime extraction too

qnkhuat committed:
update schema

qnkhuat committed:
shouldn't be able to create it in the aggregation

qnkhuat committed:
test fixes 🤨

qnkhuat committed:
miss a )

qnkhuat committed:
merge upstream

qnkhuat committed:
resolve conflicts

qnkhuat committed:
reverse some uncessary chagnes

qnkhuat committed:
simplify tests

qnkhuat committed:
clj-kondo fixes and resolve conflicts

qnkhuat committed:
Merge branch 'master' into ngoc-advanced-datetime-extraction

qnkhuat committed:
simplify tests

qnkhuat committed:
test fixes 🤨

qnkhuat committed:
datetime-extract -> temporal-extract

qnkhuat committed:
update helpers fn to supports millisecond

qnkhuat committed:
correctly check mongo version

qnkhuat committed:
resolve conflicts
